### PR TITLE
Add support for 1D MDHisto workspace to SaveASCII

### LINKED ISSUE #36646: Add support for 1D MDHisto objects in SaveASCII
**Is your feature request related to a problem? Please describe.**
WISH users want to be able to save 1D MDHisto cuts made using the non-axis aligned sliceviewer tool as an ASCII

**Describe the solution you'd like**
WISH users want `SaveASCII` to support this workspace type

**Describe alternatives you've considered**
Could have separate algorithm e.g. `Save1DMDHistoASCII` but good to keep the number of algorithms down! But this is acceptable as a solution.

**Additional context**
Here is some python code
(1) Generate a 1D MDHisto workspace
```
# create a workspace
CreateMDWorkspace(Dimensions=3, Extents='-5,5,-4,4,-3,3', Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.', Frames='HKL,HKL,HKL', SplitInto='2', SplitThreshold=50, OutputWorkspace='ws') # 2023-10-18T13:20:50.683328000 execCount: 19
FakeMDEventData(InputWorkspace='ws', UniformParams='100000', PeakParams='100000,0,0,1,0.3', RandomSeed=3873875) # 2023-10-18T13:20:50.737205000 execCount: 24
ws_cut = BinMD(InputWorkspace='ws', AxisAligned=False, BasisVector0='(0.0+1.0x 0.0 0.0), in 7.26 Ang^-1, 1.0,0.0,0.0', BasisVector1='u2, in 7.26 Ang^-1, 0.0,1.0,0.0', BasisVector2='u3, in 3.14 Ang^-1, 0.0,0.0,1.0', 
               OutputExtents='-2.5,2.5,-0.16,0.16,-0.05,0.05', OutputBins='50,1,1', NormalizeBasisVectors=False, OutputWorkspace='ws_1Dcut') # 2023-10-18T13:21:00.379543000 execCount: 26

```
(2) Save 1D MDHisto workspace `ws_1`Dcut`
```
# save it to file
dim = ws_cut.getDimension(0)
nbins = dim.getNBins()
xye = np.zeros((nbins, 3))
xye[:,1] = ws_cut.getSignalArray().flat
xye[:,2] = np.sqrt(ws_cut.getErrorSquaredArray().flat)
xye[:,0] = np.linspace(dim.getMinimum() + dim.getBinWidth()/2, dim.getMaximum() - dim.getBinWidth()/2, nbins)
# get header (using workspace history!)
alg = ws_cut.getHistory().lastAlgorithm()
header = str({prop.name: prop.valueAsStr for prop in list(alg.getProperties()) if prop.valueAsStr})
np.savetxt(r"C:\Users\xhg73778\Desktop\cut.txt", xye, fmt='%1.4e', delimiter=",", header=header)  
```

### PR DESCRIPTION
**Pull request by MohamedAlmaki**
### Description of work
This PR adds support for saving 1D MDHisto cuts using SaveASCII2

Fixes #36646. 

### To test:
(1) Generate a 1D MDHisto workspace
```
CreateMDWorkspace(Dimensions=3, Extents='-5,5,-4,4,-3,3', Names='H,K,L', Units='r.l.u.,r.l.u.,r.l.u.', Frames='HKL,HKL,HKL', SplitInto='2', SplitThreshold=50, OutputWorkspace='ws') # 2023-10-18T13:20:50.683328000 execCount: 19
FakeMDEventData(InputWorkspace='ws', UniformParams='100000', PeakParams='100000,0,0,1,0.3', RandomSeed=3873875) # 2023-10-18T13:20:50.737205000 execCount: 24
ws_cut = BinMD(InputWorkspace='ws', AxisAligned=False, BasisVector0='(0.0+1.0x 0.0 0.0), in 7.26 Ang^-1, 1.0,0.0,0.0', BasisVector1='u2, in 7.26 Ang^-1, 0.0,1.0,0.0', BasisVector2='u3, in 3.14 Ang^-1, 0.0,0.0,1.0', 
               OutputExtents='-2.5,2.5,-0.16,0.16,-0.05,0.05', OutputBins='50,1,1', NormalizeBasisVectors=False, OutputWorkspace='ws_1Dcut') # 2023-10-18T13:21:00.379543000 execCount: 26

```

(2) Save 1D MDHisto workspace `ws_1Dcut`
```
# save it to file
SaveAscii(InputWorkspace=ws_1Dcut, Filename=r"C:\Users\test.text")
```
check if the cut is saved in your preferred path

(3) Try to load the file using Load and it should be loaded as a Workspace2D

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/DataHandling/src/SaveAscii2.cpp [hunk 1]
```diff
@@ -603,4 +607,75 @@ void SaveAscii2::writeFileHeader(const std::vector<std::string> &logList, std::o
...
+  auto step = (end - start) / float(nbins - 1);
+
+  auto nPoints = mdws->getNPoints();
+  auto signal = mdws->getSignalArray();
+  auto error = mdws->getErrorSquaredArray();
+
+  Progress progress(this, 0.0, 1.0, nbins);
+
+  for (size_t i = 0; i < nPoints; ++i) {
+    file << start + step * float(i) << m_sep << signal[i] << m_sep << error[i] << "\n";
```
Comment (RichardWaiteSTFC) [inline Framework/DataHandling/src/SaveAscii2.cpp:672]: I think `SaveASCII` saves xye - so this should be sqrt(error)

File: Framework/DataHandling/test/SaveAscii2Test.h [hunk 2]
```diff
@@ -545,7 +545,7 @@ class SaveAscii2Test : public CxxTest::TestSuite {
 
     TS_ASSERT_THROWS_ANYTHING(save.execute());
 
-    // the algorithm shouldn't have written a file to disk
+    // the algorithm shouldn"t have written a file to disk
```
Comment (RichardWaiteSTFC) [inline Framework/DataHandling/test/SaveAscii2Test.h:548]: This looks like a typo to me (and I know I'm also really guilty of typos in comments...) - perhaps `'` got auto-replaced with `"` in this file? Makes the diff larger than it needs to be!

File: Framework/DataHandling/src/SaveAscii2.cpp [hunk 3]
```diff
@@ -603,4 +607,75 @@ void SaveAscii2::writeFileHeader(const std::vector<std::string> &logList, std::o
   outputFile << '\n';
 }
 
+void SaveAscii2::write1DHistoCut(const IMDHistoWorkspace_const_sptr &mdws, const std::string &filename,
+                                 bool appendToFile, bool writeHeader, int prec, bool scientific,
+                                 const std::string &comment) {
+  int count = 0;
+  size_t dimIndex = -1;
+  for (size_t i = 0; i < mdws->getNumDims(); i++) {
+    if (mdws->getDimension(i)->getNBins() > 1) {
```
Comment (RichardWaiteSTFC) [inline Framework/DataHandling/src/SaveAscii2.cpp:616]: I just found that an MD workspace has a method `getNonIntegratedDimensions()` that returns a vector - so we can check the length of that vector is 1 and also then we don't need to do this loop to get the index of the non-integrated dimension!

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
Thanks works well - just a small change to write error instead of error-squared and maybe use a method of the MD workspaces to make the code a bit more concise.

Also the hcolumn header is just the arguments of the last called algorithm (this is useful actually!) but can we also have another row along the lines of
`# Qy, Intensity, Error`
where Qy is the dimension label got from the non-integrated dimension?

File: Framework/DataHandling/test/SaveAscii2Test.h [hunk 4]
```diff
@@ -1249,6 +1286,21 @@ class SaveAscii2Test : public CxxTest::TestSuite {
...
+                                      "2", "SplitThreshold", "50", "OutputWorkspace", name.c_str());
+
+    FrameworkManager::Instance().exec("FakeMDEventData", 8, "InputWorkspace", name.c_str(), "UniformParams", "100000",
+                                      "PeakParams", "100000,0,0,1,0.3", "RandomSeed", "3873875");
+
+    FrameworkManager::Instance().exec(
+        "BinMD", 18, "InputWorkspace", name.c_str(), "AxisAligned", "0", "BasisVector0",
+        "(0.0+1.0x 0.0 0.0), in 7.26 Ang^-1, 1.0,0.0,0.0", "BasisVector1", "u2, in 7.26 Ang ^ -1, 0.0, 1.0, 0.0",
+        "BasisVector2", "u3, in 3.14 Ang ^ -1,0.0,0.0,1.0", "OutputExtents", "-2.5,2.5,-0.16,0.16,-0.05,0.05",
+        "OutputBins", "50,1,1", "NormalizeBasisVectors", "1", "OutputWorkspace", name.c_str());
```
Comment (RichardWaiteSTFC) [inline Framework/DataHandling/test/SaveAscii2Test.h:1301]: Maybe we could also change this to have `OutputBins = "1,50,1"` as a test that we're getting the non integrated dimension not the first dimension. Sorry only just spotted this!

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
Thanks for the code changes so far - looks good!
Just a small change to the unit test (sorry I didn't spot this before), and the column headers I previously requested - in python you could get the header for the Q dimension like so
```
dim = ws.getNonIntegratedDimensions()[0]
qheader = f"{dim.name} {dim.getUnits()}"
```
which for the test code gives
`u2 in 7.26 Ang^-1`

File: Framework/DataHandling/test/SaveAscii2Test.h (hunk 4 above)
Comment (MohamedAlmaki) [reply Framework/DataHandling/test/SaveAscii2Test.h:1301]: Sorry, do you mean I should add a new header for the Q dimension?

**Comment by RichardWaiteSTFC** [comment]
> Sorry, do you mean I should add a new header for the Q dimension?

As per the first review https://github.com/mantidproject/mantid/pull/36994#pullrequestreview-1928554873
It is common in these files to have a header row which provides a label for each column. If you could add a second header row (below the first one that prints the last run algorithm parameters) that'd be great thanks!
Something like
`# u2 in 7.26 Ang^-1, Signal, Error`

File: Framework/DataHandling/src/SaveAscii2.cpp [hunk 5]
```diff
@@ -603,4 +607,78 @@ void SaveAscii2::writeFileHeader(const std::vector<std::string> &logList, std::o
...
+      if (i < propsVec.size() - 1)
+        file << m_sep << " ";
+    }
+    file << " }\n";
+
+    // write columns labels
+    auto dimName = dim->getName();
+    auto dimUnit = dim->getUnits().latex();
+
+    file << dimName << " " << dimUnit << m_sep << " "
```
Comment (RichardWaiteSTFC) [inline Framework/DataHandling/src/SaveAscii2.cpp:652]: Thanks for adding to column headers, just missing a comment symbol `#` here

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
Thanks for adding the header line and changing the test to have the cut axis along the second dimension. Just one small change and then we're good!

**Review by RichardWaiteSTFC — APPROVED** [review]
Looks great thanks!

**Pull request merged**